routes/_middleware.ts

```diff
@@ -48,171 +49,179 @@ const isMonitoringRobots = (req: Request) => {
-    Deno.env.has("DECO_ANONYMOUS_CACHE")
-  ) {
-    newHeaders.set("cache-control", "public, max-age=10");
-  }
-
-  for (const flag of state?.flags ?? []) {
-    newHeaders.append(
-      DECO_MATCHER_HEADER_QS,
-      `${flag.name}=${flag.value ? 1 : 0}`,
     );
-  }
-
-  const newResponse = new Response(initialResponse.body, {
-    status: initialResponse.status,
-    headers: newHeaders,
-  });
+  },
+  async (
+    req: Request,
+    ctx: MiddlewareHandlerContext<DecoState<MiddlewareConfig, DecoSiteState>>
+  ) => {
+    if (req.method === "HEAD" && isMonitoringRobots(req)) {
+      return new Response(null, { status: 200 });
+    }
+    const url = new URL(req.url);
+    ctx.state.site = {
+      id: context.siteId,
+      name: context.site,
+    };
+
+    // FIXME (mcandeia) compatibility only.
+    if (
+      url.searchParams.has("editorData") &&
+      !url.pathname.startsWith("/live/editorData")
+    ) {
+      url.pathname = "/live/editorData";
+      url.searchParams.set("forceFresh", "");
+      return redirectTo(url);
+    }
+
+    if (url.pathname.startsWith("/_live/workbench")) {
+      url.pathname = "/live/workbench";
+      return redirectTo(url);
+    }
+
+    if (
+      !url.pathname.startsWith("/live/previews") &&
+      url.searchParams.has("pageId") &&
+      !url.searchParams.has("editorData")
+    ) {
+      return redirectToPreviewPage(url, url.searchParams.get("pageId")!);
+    }
+
+    const response = {
+      headers: new Headers(defaultHeaders),
+      status: undefined,
+    };
+    const state = ctx.state?.$live?.state ?? {};
+    const stateBag = new WeakMap();
+    state.response = response;
+    state.bag = stateBag;
+    state.flags = [];
+    Object.assign(ctx.state, state);
+    ctx.state.global = { ...(ctx.state.global ?? {}), ...state }; // compatibility mode with functions.
+
+    const shouldAllowCorsForOptions =
+      req.method === "OPTIONS" && isAdminOrLocalhost(req);
+    const initialResponse = shouldAllowCorsForOptions
+      ? new Response()
+      : await ctx.next();
+
+    // Let rendering occur — handlers are responsible for calling ctx.state.loadPage
+    if (req.headers.get("upgrade") === "websocket") {
+      return initialResponse;
+    }
+    const newHeaders = new Headers(initialResponse.headers);
+    if (
+      (url.pathname.startsWith("/live/previews") &&
+        url.searchParams.has("mode") &&
+        url.searchParams.get("mode") == "showcase") ||
+      url.pathname.startsWith("/_frsh/") ||
+      shouldAllowCorsForOptions
+    ) {
+      Object.entries(allowCorsFor(req)).map(([name, value]) => {
+        newHeaders.set(name, value);
+      });
+    }
+    response.headers.forEach((value, key) => newHeaders.append(key, value));
+    const printTimings = ctx?.state?.t?.printTimings;
+    printTimings && newHeaders.set("Server-Timing", printTimings());
+
+    if (
+      url.pathname.startsWith("/_frsh/") &&
+      [400, 404, 500].includes(initialResponse.status)
```

Comment:
I suggest you to extract the variable `status` and reuse it here to keep consistency across usages.